### PR TITLE
LDAP and ADLAP backend changes

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## Description of changes
Backend changes for LDAP and ADLDAP support

## Related issue
#227 

### How to set up LDAP to test
 1. Download (https://directory.apache.org/studio/download/download-linux.html) and setup an ldap server with apache directory studio, this guide explains it step by step to get basic users https://docs.craftercms.org/en/3.0/developers/cook-books/how-tos/setting-up-an-ldap-server-for-dev.html
 2. Start IRIDA (no ldap) and create a user account that exists in your ldap server (ex. jwick)
 3. Stop IRIDA, then configure your authentication.configuration file, below is an an example that should work with the above guide
```
# Alternative Authentication modes: ldap, adldap
# Defaults is in memory authentication with bcrypt
irida.administrative.authenitcation.mode=ldap

# LDAP config
irida.administrative.authenitcation.ldap.url=ldap://localhost:10389/
irida.administrative.authentication.ldap.base=dc=example,dc=com
irida.administrative.authenitcation.ldap.userdn=uid=admin,ou=system
irida.administrative.authenitcation.ldap.password=secret
irida.administrative.authentication.ldap.userdn_search_patterns=cn={0},ou=Users
# Optional Argument. change this to 'follow' to handle referrals. See AbstractContextSource javadocs for more info
irida.administrative.authenitcation.ldap.set_referral=ignore
```
 4. On Directory Studio, change jwick's password to something you know, like `password`
 5. Start IRIDA, and sign in with the user jwick

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [ ] Tests added (or description of how to test) for any new features.
* [ ] User documentation updated for UI or technical changes.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -156,6 +156,11 @@
 			<artifactId>spring-security-core</artifactId>
 			<version>${spring.security.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-ldap</artifactId>
+			<version>5.2.2.RELEASE</version>
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply pom.xml]: Fixed.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java [hunk 2]
```diff
@@ -76,10 +120,28 @@ protected MethodSecurityExpressionHandler createExpressionHandler() {
...
+	 * @return {@link UserDetailsContextMapper}
+	 */
+	@Bean
+	public UserDetailsContextMapper userDetailsContextMapper() {
+		return new UserDetailsContextMapper() {
+			@Override
+			public UserDetails mapUserFromContext(DirContextOperations dirContextOperations, String username, Collection<? extends GrantedAuthority> collection) {
+				// Here we could use dirContextOperations to fetch other user attributes from ldap, not needed for our use case
+				return userRepository.loadUserByUsername(username);
+			}
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java]: That's where we would redirect them to the account creation screen, if I'm understanding your question correctly.
If you're asking about the mapUserToContext, that would be a situation where a user exists in the db, but not on LDAP, which doesn't make sense for our use case.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java [hunk 3]
```diff
@@ -92,8 +154,36 @@ private AuthenticationProvider anonymousAuthenticationProvider() {
...
+		switch(authenicationMode)
+		{
+			case "ldap":
+				provider = LdapAuthenticationProvider();
+				break;
+			case "adldap":
+				provider = ActiveDirectoryLdapAuthenticationProvider();
+				break;
+			default:
+				provider = DaoAuthenticationProvider();
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java]: Fixed and added.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java [hunk 4]
```diff
@@ -92,8 +154,36 @@ private AuthenticationProvider anonymousAuthenticationProvider() {
...
+		}
+
+		return provider;
+	}
+
+	/**
+	 * Default "in memory" authentication.
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java]: Good catch, made the wording more accurate.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java [hunk 5]
```diff
@@ -92,8 +107,43 @@ private AuthenticationProvider anonymousAuthenticationProvider() {
...
 
+
+	/**
+	 * Builds and returns an {@link AuthenticationProvider} based on the irida.administrative.authentication.mode config option
+	 *
+	 * @return {@link AuthenticationProvider}
+	 */
 	@Bean
 	public AuthenticationProvider authenticationProvider() {
+		IridaLdapSecurityConfig iridaLdapSecurityConfig = new IridaLdapSecurityConfig();
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java:118]: Calling `new IridaLdapSecurityConfig()` means that Spring won't be managing this class. Instead you could try to `@Import({ IridaLdapSecurityConfig.class })` at the top of this config file.  That way it'll be managed by Spring.  You'll need to do some refactors on how this `public AuthenticationProvider authenticationProvider` method works, but that should help with the `NullPointerException` issue you were having on startup.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java]: Ok.  For now I think it's probably worth throwing an exception when the username isn't found in this case.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java (hunk 5 above)
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaApiSecurityConfig.java]: Fixed as per conversation

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaAuthenticationSecurityConfig.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    public LdapContextSource ldapContextSource() {
+        LdapContextSource ldapContextSource = new LdapContextSource();
+        ldapContextSource.setUrl(ldapUrl);
+        ldapContextSource.setBase(ldapBase);
+        ldapContextSource.setUserDn(ldapUserDn);
+        ldapContextSource.setPassword(ldapPassword);
+        ldapContextSource.setReferral(ldapSetReferral);
+        ldapContextSource.afterPropertiesSet();
+        return ldapContextSource;
+    }
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaAuthenticationSecurityConfig.java:196]: Maybe same thing with the AD LDAP provider please too.

**Review by tom114 — CHANGES_REQUESTED** [review]

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaAuthenticationSecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        return authenticationProvider;
+    }
+
+    /**
+     * This generates a ContextSource with credentials to access the LDAP server
+     *
+     * @return {@link LdapContextSource}
+     */
+    @Bean
+    public LdapContextSource ldapContextSource() {
```
[the remark this answers is not in the document]
Comment (JeffreyThiessen) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaAuthenticationSecurityConfig.java]: Fixed.

**Review by tom114 — APPROVED** [review]

**Pull request merged**